ipld/car/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use blockstore::BlockStore;
+use cid::Cid;
+use forest_encoding::from_slice;
+use serde::{Deserialize, Serialize};
+use std::io::{BufReader, Read};
+
+mod error;
+mod util;
+use crate::util::read_node;
+use error::*;
+use util::ld_read;
+
+/// CAR file header
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct CarHeader {
+    pub roots: Vec<Cid>,
+    pub version: u64,
+}
+
+impl CarHeader {
+    /// Creates a new CAR file header
+    pub fn new(roots: Vec<Cid>, version: u64) -> Self {
+        Self { roots, version }
+    }
+}
+
+/// Reads CAR files that are in a BufReader
+pub struct CarReader<R> {
+    pub buf_reader: BufReader<R>,
+    pub header: CarHeader,
+}
+
+impl<R> CarReader<R>
+where
+    R: std::io::Read,
```

Comment:
nit: move this to the constructor, no need to constrain the whole impl (future proof)